File: 0056-merge-intervals/0056-merge-intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:        
        result = []
        intervals.sort()
        current_interval = intervals[0]

        for i in range(1, len(intervals)):
            if current_interval[1] >= intervals[i][0]:
                merged_interval = [current_interval[0], max(intervals[i][1], current_interval[1])]
                current_interval = merged_interval
            else:
                result.append(current_interval)
                current_interval = intervals[i]

        result.append(current_interval)
        
        return result

File: 0056-merge-intervals/0056-merge-intervals.java
class Solution {
    public int[][] merge(int[][] intervals) {
        if (intervals.length <= 1) return intervals;

        int[] last_merged;
        List<int[]> res = new ArrayList<>();

        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));

        last_merged = intervals[0];
        for (int i=1; i<intervals.length; i++) {
            int r1 = last_merged[1];
            int l2 = intervals[i][0];

            if (r1 >= l2) {
                last_merged[1] = Math.max(r1, intervals[i][1]);
            } else {
                res.add(last_merged);
                last_merged = intervals[i];
            }
        }

        res.add(last_merged);

        return res.toArray(new int[res.size()][]);
    }
}